From dfca28d0d8d543ea37b584d9639be13815445387 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:05:16 -0400
Subject: [usb] ch341: clean up line-status handling

Message-id: <1410887463-20674-245-git-send-email-dzickus@redhat.com>
Patchwork-id: 95169
O-Subject: [RHEL7 PATCH 244/591] USB: ch341: clean up line-status handling
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit b770081f88b75212d61a63a84274e491eb54b25a
Author: Johan Hovold <jhovold@gmail.com>
Date:   Thu Jan 2 22:49:29 2014 +0100

    USB: ch341: clean up line-status handling

    Clean up line-status handling.

    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/serial/ch341.c b/drivers/usb/serial/ch341.c
index f647dbd..9dd94a7 100644
--- a/drivers/usb/serial/ch341.c
+++ b/drivers/usb/serial/ch341.c
@@ -444,24 +444,30 @@ static void ch341_update_line_status(struct usb_serial_port *port,
      unsigned char *data, size_t len)
 {
  struct ch341_private *priv = usb_get_serial_port_data(port);
+ struct tty_struct *tty;
  unsigned long flags;
- u8 prev_line_status = priv->line_status;
+ u8 status;
+ u8 delta;
 
  if (len < 4)
   return;
 
+ status = ~data[2] & CH341_BITS_MODEM_STAT;
+
  spin_lock_irqsave(&priv->lock, flags);
- priv->line_status = (~(data[2])) & CH341_BITS_MODEM_STAT;
- if ((data[1] & CH341_MULT_STAT))
+ delta = status ^ priv->line_status;
+ priv->line_status = status;
+ if (data[1] & CH341_MULT_STAT)
   priv->multi_status_change = 1;
  spin_unlock_irqrestore(&priv->lock, flags);
 
- if ((priv->line_status ^ prev_line_status) & CH341_BIT_DCD) {
-  struct tty_struct *tty = tty_port_tty_get(&port->port);
-  if (tty)
+ if (delta & CH341_BIT_DCD) {
+  tty = tty_port_tty_get(&port->port);
+  if (tty) {
    usb_serial_handle_dcd_change(port, tty,
-     priv->line_status & CH341_BIT_DCD);
-  tty_kref_put(tty);
+      status & CH341_BIT_DCD);
+   tty_kref_put(tty);
+  }
  }
 
  wake_up_interruptible(&port->port.delta_msr_wait);
-- 
1.7.1